library/core/registry/src/main/java/org/quiltmc/qsl/registry/impl/event/RegistryMonitorImpl.java

```diff
@@ -0,0 +1,53 @@
+package org.quiltmc.qsl.registry.impl.event;
+
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.qsl.registry.api.event.RegistryEntryPredicate;
+import org.quiltmc.qsl.registry.api.event.RegistryEvents;
+import org.quiltmc.qsl.registry.api.event.RegistryMonitor;
+
+public class RegistryMonitorImpl<V> implements RegistryMonitor<V> {
+	private final Registry<V> registry;
+	private @Nullable RegistryEntryPredicate<V> filter = null;
+
+	public RegistryMonitorImpl(Registry<V> registry) {
+		this.registry = registry;
+	}
+
+	@Override
+	public RegistryMonitor<V> filter(RegistryEntryPredicate<V> filter) {
+		this.filter = this.filter == null ? filter : this.filter.and(filter);
+		return this;
+	}
+
+	@Override
+	public void forAll(RegistryEvents.EntryAdded<V> callback) {
+		registry.forEach(entry -> {
```

Comment:
I swear there's a better way to do this, it involves accessors but should reduce the number of lookups needed, which should be considered especially with large registries in mind.

Reply:
Ideally we would register the callback and then iterate the registry here.

Reply:
To allow chaining?
Mmhh, unsure, since it could also allow easy to miss recursion.